intellij-plugin-structure/structure-intellij/src/main/java/com/jetbrains/plugin/structure/intellij/verifiers/PluginUntilBuildVerifier.kt

```diff
@@ -27,43 +33,68 @@ class PluginUntilBuildVerifier {
     if (untilBuildParsed == null) {
       registerProblem(InvalidUntilBuild(descriptorPath, untilBuild))
     } else {
-      if (untilBuildParsed.baselineVersion >= 999) {
-        registerProblem(InvalidUntilBuild(descriptorPath, untilBuild, untilBuildParsed))
-      } else if (untilBuildParsed.baselineVersion > 400) {
-        registerProblem(SuspiciousUntilBuild(untilBuild))
-      }
+      verifyBaseLineVersion(untilBuildParsed, untilBuild, descriptorPath, problemRegistrar)
       if (untilBuildParsed.productCode.isNotEmpty()) {
         registerProblem(ProductCodePrefixInBuild(descriptorPath))
       }
     }
   }
 
-  private fun isSpecialSingleComponent(untilBuild: String): Boolean {
-    return BUILD_NUMBER == untilBuild || SNAPSHOT == untilBuild
+  private fun verifyBaseLineVersion(untilBuild: IdeVersion,
+                                    untilBuildValue: String,
+                                    descriptorPath: String,
+                                    problemRegistrar: ProblemRegistrar) = with(problemRegistrar) {
+    untilBuild.baselineVersion.let {
+      verifyBaseline(it, untilBuildValue, untilBuild, descriptorPath, problemRegistrar)
+    }
+  }
+
+  private fun verifyBaseline(baseline: Int,
+                             untilBuildValue: String,
+                             untilBuild: IdeVersion?,
+                             descriptorPath: String,
+                             problemRegistrar: ProblemRegistrar) = with(problemRegistrar) {
+    if (baseline >= 999) {
+      registerProblem(InvalidUntilBuild(descriptorPath, untilBuildValue, untilBuild))
+    } else if (baseline >= SUSPICIOUS_BASELINE_LOWER_BOUND) {
+      registerProblem(SuspiciousUntilBuild(untilBuildValue))
+    } else {
+      verifyInThreeReleasesPerYear(untilBuildValue, baselineVersion = baseline, problemRegistrar)
+    }
   }
 
   private fun verifySingleComponentUntilBuild(untilBuild: String,
                                               descriptorPath: String,
                                               problemRegistrar: ProblemRegistrar) {
     try {
       val untilBuildNumber = untilBuild.toInt()
-      if (untilBuildNumber >= 400) {
-        problemRegistrar.registerProblem(InvalidUntilBuild(descriptorPath, untilBuild))
-      } else if (untilBuildNumber >= 300) {
-        problemRegistrar.registerProblem(SuspiciousUntilBuild(untilBuild))
-      }
+      verifyBaseline(untilBuildNumber, untilBuild, untilBuild = null, descriptorPath, problemRegistrar)
     } catch (e: NumberFormatException) {
       problemRegistrar.registerProblem(InvalidUntilBuild(descriptorPath, untilBuild))
     }
   }
 
+  private fun verifyInThreeReleasesPerYear(untilBuildValue: String,
```

Comment:
this may or may not be true in some (distant) future

Reply:
Then we need to decide if
1. validate and change it when and if in that (distant) future the release cycle will change
2. or not validate this rule at all.

Note that this is as per docs

> Any of 999.*, 234.* (maximum is 233.*) and 223.9999 are invalid for `until-build`.

Please advice.

Reply:
We can keep the check for now, but one day we might need to dynamically adjust it.